velox/row/CompactRow.cpp

```diff
@@ -17,6 +17,153 @@
+    std::vector<size_t>& /* unused */) {
+  for (auto i = 0; i < rows.size(); ++i) {
+    bits::setBit(nulls[i], childIdx, true);
+  }
+}
+
+template <>
+void serializeTyped<TypeKind::BOOLEAN>(
+    const raw_vector<vector_size_t>& rows,
+    uint32_t childIdx,
+    DecodedVector& decoded,
+    size_t valueBytes,
+    const raw_vector<uint8_t*>& nulls,
+    char* buffer,
+    std::vector<size_t>& offsets) {
+  const auto mayHaveNulls = decoded.mayHaveNulls();
+
+  for (auto i = 0; i < rows.size(); ++i) {
+    if (mayHaveNulls && decoded.isNullAt(rows[i])) {
+      bits::setBit(nulls[i], childIdx, true);
+    } else {
+      // Write fixed-width value.
+      reinterpret_cast<bool*>(buffer)[offsets[i]] =
+          decoded.valueAt<bool>(rows[i]);
+    }
+    offsets[i] += valueBytes;
+  }
+}
+
+template <>
+void serializeTyped<TypeKind::TIMESTAMP>(
+    const raw_vector<vector_size_t>& rows,
+    uint32_t childIdx,
+    DecodedVector& decoded,
+    size_t valueBytes,
+    const raw_vector<uint8_t*>& nulls,
+    char* buffer,
+    std::vector<size_t>& offsets) {
+  const auto mayHaveNulls = decoded.mayHaveNulls();
+  const auto* rawData = decoded.data<Timestamp>();
+
+  for (auto i = 0; i < rows.size(); ++i) {
+    if (mayHaveNulls && decoded.isNullAt(rows[i])) {
+      bits::setBit(nulls[i], childIdx, true);
+    } else {
+      // Write fixed-width value.
+      auto micros = rawData[rows[i]].toMicros();
+      memcpy(buffer + offsets[i], &micros, sizeof(int64_t));
+    }
+    offsets[i] += valueBytes;
+  }
+}
+
+void serializeVarcharAndVarBinary(
+    const raw_vector<vector_size_t>& rows,
+    uint32_t childIdx,
+    DecodedVector& decoded,
+    const raw_vector<uint8_t*>& nulls,
+    char* buffer,
+    std::vector<size_t>& offsets) {
+  const auto mayHaveNulls = decoded.mayHaveNulls();
+
+  for (auto i = 0; i < rows.size(); ++i) {
+    if (mayHaveNulls && decoded.isNullAt(rows[i])) {
+      bits::setBit(nulls[i], childIdx, true);
+    } else {
+      auto value = decoded.valueAt<StringView>(rows[i]);
+      writeInt32(buffer + offsets[i], value.size());
+      if (!value.empty()) {
+        memcpy(buffer + offsets[i] + kSizeBytes, value.data(), value.size());
+      }
+      offsets[i] += kSizeBytes + value.size();
+    }
+  }
+}
+
+template <>
+void serializeTyped<TypeKind::VARCHAR>(
+    const raw_vector<vector_size_t>& rows,
+    uint32_t childIdx,
+    DecodedVector& decoded,
+    size_t valueBytes,
+    const raw_vector<uint8_t*>& nulls,
+    char* buffer,
+    std::vector<size_t>& offsets) {
+  serializeVarcharAndVarBinary(rows, childIdx, decoded, nulls, buffer, offsets);
+}
+
+template <>
+void serializeTyped<TypeKind::VARBINARY>(
+    const raw_vector<vector_size_t>& rows,
+    uint32_t childIdx,
+    DecodedVector& decoded,
+    size_t valueBytes,
+    const raw_vector<uint8_t*>& nulls,
+    char* buffer,
+    std::vector<size_t>& offsets) {
+  serializeVarcharAndVarBinary(rows, childIdx, decoded, nulls, buffer, offsets);
```

Comment:
Does it make sense to use std::enable_if for VARCHAR and VARBINARY and move the implementation of `serializeVarcharAndVarBinary` here?

```C++
template <TypeKind kind>
typename std::enable_if<kind == TypeKind::VARCHAR || kind == TypeKind::VARBINARY, void>::type
serializeTyped(
```

Reply:
If we add this to the varchar/varbinary, the default template function would become cumbersome with the `std::enable_if<...>` condition. To simplify the code, we can move the implementation into `serializeTyped<VARCHAR>` and call it in `serializeTyped<VARBINARY>`.